packages/vm/src/parsed_wasm.rs

```diff
@@ -147,4 +176,33 @@ impl<'a> ParsedWasm<'a> {
 
         Ok(this)
     }
+
+    /// Perform the expensive operation of validating each function body
+    ///
+    /// Note: This function caches the output of this function into the field `funcs_to_validate` so repeated invocations are cheap.
+    pub fn validate_funcs(&mut self) -> VmResult<()> {
+        match self.func_validator {
+            FunctionValidator::Pending(OpaqueDebug(ref mut funcs)) => {
+                let result = mem::take(funcs) // This is fine since `Vec::default()` doesn't allocate
+                    .into_par_iter()
+                    .try_for_each(|(func, body)| {
+                        // Reusing the allocations between validations only results in an ~6% performance improvement
+                        // The parallelization is blowing this out of the water by a magnitude of 5x
+                        // Especially when combining this with a high-performance allocator, the missing buffer reuse will be pretty much irrelevant
+                        let mut validator = func.into_validator(Default::default());
+                        validator.validate(&body)?;
```

Comment:
Here we need to ensure we don't get different results depending on the pralllelization. I.e. every node executing this must always get the same result. Is `try_for_each` taking care of that?

Reply:
`try_for_each` is using a call to `reduce` internally. 

Since the error the parser produces (`wasmparser::BinaryReaderError`) is rather opaque (no `PartialEq`, no `PartialOrd`, just a message and an offset), I think we can treat the reduce operation here as associative, meaning the order in which the values for reduce are produced isn't really relevant.

But we can of course simply remove the `par_iter` and the `rayon` dependency and perform it sequentially.

Reply:
The docs of `try_for_each` specify that it is not specified which error we get in case of multiple errors.
We currently use the `Display` impl of the `BinaryReaderError` when [converting to our `VmError` type](https://github.com/CosmWasm/cosmwasm/blob/a934ea73b1d5a143a433d1e9b6a70cb35e9210e5/packages/vm/src/errors/vm_error.rs#L267), so this is non-deterministic. I see two possible options here:
- remove rayon again and do this sequentially
- change our error conversion to not include the original error (might be more annoying for users to debug)

Reply:
For now I simply removed `rayon` again since that function is not even relevant for the `analyze` call itself. It would, at best, improve the performance of `Cache::save_wasm`.

Reply:
Thanks, sounds good.

I was wondering if we check all in parallel, collect the results and return the first error. But we want early exits to protect against malicious inputs creating unnecessary load on the machines.



> that function is not even relevant for the `analyze` call itself. It would, at best, improve the performance of `Cache::save_wasm`.

Good point. And `save_wasm` is slow anyways since it compiled the Wasm to the native machine code. This is executed way less often than `analyze`.

Reply:
> I was wondering if we check all in parallel, collect the results and return the first error. But we want early exits to protect against malicious inputs creating unnecessary load on the machines.

I *don't think* it does. Since it seems like it's running reduce internally, I'm assuming it just runs logic akin to `if acc.is_err() { acc } else { result }` without early exits.

Reply:
I think this is not critical to optimize, so let's keep it simple. This is only executed together with the compilation step now, which is expected to not happen often. `analyze` got much faster here which is what we need.